code/game/objects/effects/portals.dm

```diff
@@ -1,69 +1,143 @@
 
+/proc/create_portal_pair(turf/source, turf/destination, _creator = null, _lifespan = 300, accuracy = 0)
+	if(!istype(source) || !istype(destination))
+		return
+	var/turf/actual_destination = get_teleport_turf(destination, accuracy)
+	var/obj/effect/portal/P1 = new(source, _creator, _lifespan, null, FALSE)
+	var/obj/effect/portal/P2 = new(actual_destination, _creator, _lifespan, P1, TRUE)
+	P1.link_portal(P2)
+	P1.hardlinked = TRUE
+	return list(P1, P2)
+
 /obj/effect/portal
 	name = "portal"
 	desc = "Looks unstable. Best to test it with the clown."
 	icon = 'icons/obj/stationobjs.dmi'
 	icon_state = "portal"
-	density = 1
-	var/obj/item/target = null
-	var/creator = null
-	anchored = 1
-	var/precision = 1 // how close to the portal you will teleport. 0 = on the portal, 1 = adjacent
+	anchored = TRUE
 	var/mech_sized = FALSE
+	var/obj/effect/portal/linked
+	var/hardlinked = TRUE			//Requires a linked portal at all times. Destroy if there's no linked portal, if there is destroy it when this one is deleted.
+	var/creator
+	var/turf/hard_target			//For when a portal needs a hard target and isn't to be linked.
+	var/atmos_link = FALSE			//Link source/destination atmos.
+	var/turf/open/atmos_source		//Atmos link source
+	var/turf/open/atmos_destination	//Atmos link destination
 
-/obj/effect/portal/Bumped(mob/M as mob|obj)
-	teleport(M)
+/obj/effect/portal/attackby(obj/item/weapon/W, mob/user, params)
+	if(user && Adjacent(user))
+		user.forceMove(get_turf(src))
+		teleport(user)
+
+/obj/effect/portal/make_frozen_visual()
+	return
+
+/obj/effect/portal/Crossed(atom/movable/AM, oldloc)
+	if(get_turf(oldloc) == get_turf(linked))
+		return ..()
+	if(!teleport(AM))
+		return ..()
 
 /obj/effect/portal/attack_tk(mob/user)
 	return
 
 /obj/effect/portal/attack_hand(mob/user)
 	if(Adjacent(user))
+		user.forceMove(get_turf(src))
 		teleport(user)
 
-/obj/effect/portal/attackby(obj/item/weapon/W, mob/user, params)
-	if(user && Adjacent(user))
-		teleport(user)
+/obj/effect/portal/Initialize(mapload, _creator, _lifespan = 300, obj/effect/portal/_linked = null, automatic_link = TRUE, hard_target_override = null, atmos_link_override = null)
+	. = ..()
+	GLOB.portals += src
+	if(!istype(_linked) && automatic_link)
+		return INITIALIZE_HINT_QDEL
+	if(_lifespan > 0)
+		QDEL_IN(src, _lifespan)
+	if(!isnull(atmos_link_override))
+		atmos_link = atmos_link_override
+	link_portal(_linked)
+	hardlinked = automatic_link
+	creator = _creator
 
-/obj/effect/portal/make_frozen_visual()
-	return
+/obj/effect/portal/proc/link_portal(obj/effect/portal/newlink)
+	linked = newlink
+	if(atmos_link)
+		unlink_atmos()
+		link_atmos()
 
-/obj/effect/portal/New(loc, turf/target, creator=null, lifespan=300)
-	..()
-	GLOB.portals += src
-	src.target = target
-	src.creator = creator
+/obj/effect/portal/proc/link_atmos()
+	if(atmos_source || atmos_destination)
+		unlink_atmos()

```

Comment:
This is doubly redundant since you already call this in the proc that calls this one, you also check for some shit for some realon while there you just call it without care
Remove it all